xentop/xentop.go

```diff
@@ -0,0 +1,193 @@
+ * Types
+ **********************************************************************************************************************/
+
+// SystemInfo domain info.
+type SystemInfo struct {
+	Name               string  `field:"NAME"`
+	State              string  `field:"STATE"`
+	CpuTime            int64   `field:"CPU(sec)"` //nolint:stylecheck
+	CpuFraction        float32 `field:"CPU(%)"`   //nolint:stylecheck
+	Memory             int64   `field:"MEM(k)"`
+	MaxMemory          int64   `field:"MAXMEM(k)"`
+	MemoryFraction     float32 `field:"MEM(%)"`
+	MaxMemoryFraction  float32 `field:"MAXMEM(%)"`
+	VirtualCpus        int64   `field:"VCPUS"`
+	NetworkInterfaces  int64   `field:"NETS"`
+	NetworkTx          int64   `field:"NETTX(k)"`
+	NetworkRx          int64   `field:"NETRX(k)"`
+	VirtualDisks       int64   `field:"VBDS"`
+	DiskBlockedIO      int64   `field:"VBD_OO"`
+	DiskReadOps        int64   `field:"VBD_RD"`
+	DiskWriteOps       int64   `field:"VBD_WR"`
+	DiskSectorsRead    int64   `field:"VBD_RSECT"`
+	DiskSectorsWritten int64   `field:"VBD_WSECT"`
+	SSID               int64   `field:"SSID"`
+}
+
+type ShellCommand interface {
+	CombinedOutput() ([]byte, error)
+}
+
+type execShellCommand struct {
+	*exec.Cmd
+}
+
+/***********************************************************************************************************************
+ * Variable
+ **********************************************************************************************************************/
+
+// ExecContext global variable are used to be able to mocking the functionality of xentop in tests.
+var ExecContext = newExecShellCommander //nolint:gochecknoglobals
+
+/***********************************************************************************************************************
+ * Public
+ **********************************************************************************************************************/
+
+// GetSystemInfo gets system information from xentop.
+func GetSystemInfos() (map[string]SystemInfo, error) {
+	output, err := ExecContext("xentop", "-b", "-f", "-i").CombinedOutput()
+	if err != nil {
+		return nil, aoserrors.Wrap(err)
+	}
+
+	xentopLines := strings.Split(string(output), "\n")
+	if len(xentopLines) < 1 {
+		return nil, aoserrors.New("xentop output empty")
+	}
+
+	header := strings.Fields(xentopLines[0])
+	if len(header) < 1 || header[0] != "NAME" {
+		return nil, aoserrors.New("unexpected xentop header")
+	}
+
+	systemInfos := make(map[string]SystemInfo)
+
+	for _, line := range xentopLines[1:] {
+		line = strings.ReplaceAll(line, "no limit", "no-limit")
+
+		info := strings.Fields(line)
+		if len(info) != len(header) {
+			return nil, aoserrors.New("unexpected xentop data")
+		}
+
+		systemInfo, err := fillDomainInfo(header, info)
+		if err != nil {
+			return nil, err
+		}
+
+		// Ignore parse system information if xentop show extremely high CPU time and CPU usage
```

Comment:
Why we should ignore it? It is xentop problem or some system problem.

Reply:
Agree. Remove it